WordPress/WordPressTest/ContextManagerTests.swift

```diff
@@ -141,6 +141,50 @@ class ContextManagerTests: XCTestCase {
         }
     }
 
+    func testSaveDerivedContextWithChangesInMainContext() throws {
+        let contextManager = ContextManagerMock()
+        let derivedContext = contextManager.newDerivedContext()
+
+        let firstUserSaved = self.expectation(description: "First user is saved")
+        derivedContext.perform {
+            let account = WPAccount(context: derivedContext)
+            account.userID = 1
+            account.username = "First User"
+            contextManager.saveContextAndWait(derivedContext)
+            firstUserSaved.fulfill()
+        }
+
+        let findFirstUser: () throws -> WPAccount? = {
+            let firstUserQuery = NSFetchRequest<WPAccount>(entityName: "Account")
+            firstUserQuery.predicate = NSPredicate(format: "userID = 1")
+            return try contextManager.mainContext.fetch(firstUserQuery).first
+        }
+
+        wait(for: [firstUserSaved], timeout: 0.1)
+        try XCTAssertEqual(findFirstUser()?.username, "First User")
+
+        // Change first user's user name
+        try findFirstUser()?.username = "First User (Updated)"
+
+        // Save another user
+        let secondUserSaved = self.expectation(description: "Second user is saved")
+        derivedContext.perform {
+            let account = WPAccount(context: derivedContext)
+            account.userID = 2
+            account.username = "Second account"
+            contextManager.saveContextAndWait(derivedContext)
+            secondUserSaved.fulfill()
+        }
+
+        wait(for: [secondUserSaved], timeout: 0.1)
+
+        // Discard the username change that's made above
+        contextManager.mainContext.reset()
+
+        XCTExpectFailure("Known issue: the mainContext is saved along with the `ContextManager.save` functions")
+        try XCTAssertEqual(findFirstUser()?.username, "First User")
```

Comment:
Have you considered using Nimble for the test, to make the syntax for the async expectations more lightweight?

```swift
        let contextManager = ContextManagerMock()
        let derivedContext = contextManager.newDerivedContext()
        
        derivedContext.perform {
            let account = WPAccount(context: derivedContext)
            account.userID = 1
            account.username = "First User"
            contextManager.saveContextAndWait(derivedContext)
        }
        
        let findFirstUser: () throws -> WPAccount? = {
            let firstUserQuery = NSFetchRequest<WPAccount>(entityName: "Account")
            firstUserQuery.predicate = NSPredicate(format: "userID = 1")
            return try contextManager.mainContext.fetch(firstUserQuery).first
        }
        expect(try findFirstUser()?.username).toEventually(equal("First User"))

        // Change first user's user name
        try findFirstUser()?.username = "First User (Updated)"

        // Save another user
        waitUntil { done in
            derivedContext.perform {
                let account = WPAccount(context: derivedContext)
                account.userID = 2
                account.username = "Second account"
                contextManager.saveContextAndWait(derivedContext)
                done()
            }
        }

        // Discard the username change that's made above
        contextManager.mainContext.reset()

        XCTExpectFailure("Known issue: the mainContext is saved along with the `ContextManager.save` functions")
        expect(try findFirstUser()?.username) == "First User"
```

I tried this locally and it passed, or rather failed as expected:

<img width="1992" alt="image" src="https://user-images.githubusercontent.com/1218433/184284802-a8b64b0b-1cac-43d5-8dcf-62533b6d4210.png">

We don't have any guideline on whether to use Nimble or not. I think it has many syntax benefits and it might perform better for some async expectations.

Reply:
👍 This does look nicer than creating and fulfilling expectations. Please feel free to push up your changes (may as well use the cleaner code considering you've already implemented it).

Reply:
Done [wordpress-mobile/WordPress-iOS@`1210edd` (#19180)](https://github.com/wordpress-mobile/WordPress-iOS/pull/19180/commits/1210edd5494684fd5502672b25a749bf8796cab0)